RecoPixelVertexing/PixelTrackFitting/src/PixelTrackReconstructionGPU.cc

```diff
@@ -0,0 +1,197 @@
+{
+  edm::InputTag filterTag = cfg.getParameter<edm::InputTag>("Filter");
+  if(filterTag.label() != "") {
+    theFilterToken = iC.consumes<PixelTrackFilter>(filterTag);
+  }
+}
+
+PixelTrackReconstructionGPU::~PixelTrackReconstructionGPU()
+{
+}
+
+void PixelTrackReconstructionGPU::fillDescriptions(edm::ParameterSetDescription& desc) {
+  desc.add<edm::InputTag>("SeedingHitSets", edm::InputTag("pixelTracksHitTriplets"));
+  desc.add<edm::InputTag>("Fitter", edm::InputTag("pixelFitterByHelixProjections"));
+  desc.add<edm::InputTag>("Filter", edm::InputTag("pixelTrackFilterByKinematics"));
+  desc.add<std::string>("Cleaner", "pixelTrackCleanerBySharedHits");
+}
+
+void PixelTrackReconstructionGPU::run(TracksWithTTRHs& tracks,
+    edm::Event& ev, const edm::EventSetup& es)
+{
+  edm::ESHandle<MagneticField> fieldESH;
+  es.get<IdealMagneticFieldRecord>().get(fieldESH);
+
+ edm::Handle<RegionsSeedingHitSets> hhitSets;
+  ev.getByToken(theHitSetsToken, hhitSets);
+  const auto& hitSets = *hhitSets;
+
+  const PixelTrackFilter *filter = nullptr;
+  if(!theFilterToken.isUninitialized()) {
+    edm::Handle<PixelTrackFilter> hfilter;
+    ev.getByToken(theFilterToken, hfilter);
+    filter = hfilter.product();
+  }
+
+  float bField = 1 / PixelRecoUtilities::fieldInInvGev(es);
+
+  std::vector<float> hits_and_covariances;
+  float * hits_and_covariancesGPU = nullptr;
+  std::vector<Rfit::helix_fit> helix_fit_results;
+  Rfit::helix_fit * helix_fit_resultsGPU = nullptr;
+
+  const int points_in_seed = 4;
+  // We use 3 floats for GlobalPosition and 9 floats for GlobalError (that's what is used by the Riemann fit).
+  // TODO: optimise dimensions eploiting matrix symmetries
+  // Assume a safe maximum of 3K seeds: it will dynamically grow, if needed.
+  int total_seeds = 0;
+  hits_and_covariances.reserve(sizeof(float)*3000*(points_in_seed*12));
+  for (auto const & regionHitSets : hitSets) {
+    const TrackingRegion& region = regionHitSets.region();
+    for (auto const & tuplet : regionHitSets) {
+      for (unsigned int iHit = 0; iHit < tuplet.size(); ++iHit) {
+        auto const& recHit = tuplet[iHit];
+        auto point = GlobalPoint(recHit->globalPosition().basicVector() - region.origin().basicVector());
+        auto errors = recHit->globalPositionError().matrix4D();
+        hits_and_covariances.push_back(point.x());
+        hits_and_covariances.push_back(point.y());
+        hits_and_covariances.push_back(point.z());
+        for (auto j = 0; j < 3; ++j) {
+          for (auto l = 0; l < 3; ++l) {
+            hits_and_covariances.push_back(errors(j, l));
+          }
+        }
+      }
+      total_seeds++;
+    }
+  }
+
+  // We pretend to have one fit for every seed
+  helix_fit_results.resize(total_seeds);
+
+  cudaCheck(cudaMalloc((void**)&hits_and_covariancesGPU, sizeof(float)*hits_and_covariances.size()));
+  cudaCheck(cudaMalloc((void**)&helix_fit_resultsGPU, sizeof(Rfit::helix_fit)*helix_fit_results.size()));
```

Comment:
[longer term] All other producers (raw2cluster, rechits, CA) will do their memory allocations at the beginning of the job. Would it make sense to the same here?